core/rmutex.c

```diff
@@ -0,0 +1,162 @@
+        /* mutex is already held
+         *
+         * Case 1: the current thread holds the mutex
+         *     Invariant 1: holds
+         *     rmutex->owner == thread_getpid()
+         *
+         * Case 2: another thread holds the mutex
+         *     Invariant 1: holds
+         *     rmutex->owner != thread_getpid()
+         *
+         * Note for Case 2:
+         *
+         *     The foreign thread with PID forgeign_pid might change
+         *     rmutex->owner.
+         *
+         *     a) either from KERNEL_PID_UNDEF to any forgeign PID if
+         *     it is currently executing rmutex_lock or rmutex_trylock
+         *     even if this happens multiple times with different PIDs
+         *     the outcome of the next if statement will not be
+         *     influenced
+         *
+         *     b) either from forgeign_pid to KERNEL_PID_UNDEF if
+         *     it is currently executing rmutex_unlock
+         *
+         *     but this does not bother the invariant
+         */
+
+        /* ensure that owner is read only once */
+        owner = *((volatile kernel_pid_t *)(&rmutex->owner));
+        DEBUG("rmutex %" PRIi16" : mutex held by %" PRIi16" \n", thread_getpid(), owner);
+
+        /* Case 2 : Another Thread hold the mutex */
+        if ( owner != thread_getpid() ) {
+            /* wait for the mutex */
+            DEBUG("rmutex %" PRIi16" : locking mutex\n", thread_getpid());
+
+            mutex_lock(&rmutex->mutex);
+        }
+        /* Case 1 : We already hold the mutex */
+        else {
+            assert(rmutex->refcount>0);
+        }
+
+        /* fall through */
+    case 1:
+
+        DEBUG("rmutex %" PRIi16" : I am now holding the mutex\n", thread_getpid());
+
+        /* We are holding the recursive mutex */
+
+        DEBUG("rmutex %" PRIi16" : settting the owner\n", thread_getpid());
+
+        /* ensure that owner is written only once */
+        *((volatile kernel_pid_t *)(&rmutex->owner)) = thread_getpid();
+
+        DEBUG("rmutex %" PRIi16" : increasing refs\n", thread_getpid());
+
+        /* increase the refcount */
+        rmutex->refcount++;
+    }
+}
+
+int rmutex_trylock(rmutex_t *rmutex)
+{
+    kernel_pid_t owner;
+    /* try to lock the mutex */
+    switch (mutex_trylock(&rmutex->mutex)) {
+    case 0:
+        /* ensure that owner is read only once */
+        owner = *((volatile uint16_t *)(&rmutex->owner));
+
+        /* Case 2 : Another Thread hold the mutex */
+        if ( owner != thread_getpid() ) {
+            /* wait for the mutex */
+            return 0;
+        }
+        /* Case 1 : We already hold the mutex */
+        else {
+            assert(rmutex->refcount>0);
+        }
+
+        /* fall through */
+    case 1:
+        /* we are holding the recursive mutex */
+
+        /* ensure that owner is written only once */
+        *((volatile uint16_t *)(&rmutex->owner)) = thread_getpid();
```

Comment:
I know it is not pretty.

The point is that i think it is necessary to read/write owner in an atomic manner, because it could be modified by any other thread. The implementation could break if the compiler, for any reason, decides to split read or write of owner into 2 single byte loads/stores.

Linux uses the ACCESS_ONCE(x) macro which RIOT does not support.

Do you have an idea to force the compiler to load/store the variable with a single store/load operation?